File: controller/routes/system.py
"""System endpoints: health, status, config, gpus."""

from __future__ import annotations

import asyncio

import httpx
from fastapi import APIRouter

from .. import __version__
from ..config import settings
from ..gpu import get_gpu_info
from ..models import (
    HealthResponse, SystemConfigResponse, ServiceInfo, SystemConfig, EnvironmentInfo
)
from ..process import find_inference_process

router = APIRouter(tags=["System"])


@router.get("/health", response_model=HealthResponse)
async def health():
    """Health check."""
    current = find_inference_process(settings.inference_port)
    inference_ready = False

    if current:
        try:
            async with httpx.AsyncClient(timeout=5) as client:
                r = await client.get(f"http://localhost:{settings.inference_port}/health")
                inference_ready = r.status_code == 200
        except Exception:
            pass

    return HealthResponse(
        status="ok",
        version=__version__,
        inference_ready=inference_ready,
        backend_reachable=inference_ready,
        running_model=current.served_model_name or current.model_path if current else None,
    )


# Global launching state (shared with lifecycle routes)
_launching_recipe_id: str | None = None


def get_launching_recipe_id() -> str | None:
    """Get the currently launching recipe ID."""
    return _launching_recipe_id


def set_launching_recipe_id(recipe_id: str | None) -> None:
    """Set the currently launching recipe ID."""
    global _launching_recipe_id
    _launching_recipe_id = recipe_id


@router.get("/status")
async def status():
    """Detailed status including launch-in-progress info."""
    current = find_inference_process(settings.inference_port)
    return {
        "running": current is not None,
        "process": current.model_dump() if current else None,
        "inference_port": settings.inference_port,
        "launching": _launching_recipe_id,
    }


@router.get("/gpus")
async def gpus():
    """Get GPU information."""
    gpu_list = get_gpu_info()
    return {
        "count": len(gpu_list),
        "gpus": [gpu.model_dump() for gpu in gpu_list],
    }


@router.get("/config", response_model=SystemConfigResponse)
async def get_config():
    """Get system configuration and service topology."""
    import os
    import socket

    async def check_service(host: str, port: int, timeout: float = 1.0) -> bool:
        try:
            _, _, _ = await asyncio.get_event_loop().getaddrinfo(host, port)
            reader, writer = await asyncio.wait_for(
                asyncio.open_connection(host, port),
                timeout=timeout
            )
            writer.close()
            await writer.wait_closed()
            return True
        except Exception:
            return False

    hostname = socket.gethostname()
    services = []

    # Controller service
    services.append(ServiceInfo(
        name="Controller",
        port=settings.port,
        internal_port=settings.port,
        protocol="http",
        status="running",
        description="FastAPI model lifecycle manager"
    ))

    # Inference backend (vLLM/SGLang)
    inference_status = "unknown"
    try:
        current = find_inference_process(settings.inference_port)
        if current:
            async with httpx.AsyncClient(timeout=2) as client:
                r = await client.get(f"http://localhost:{settings.inference_port}/health")
                inference_status = "running" if r.status_code == 200 else "error"
        else:
            inference_status = "stopped"
    except Exception:
        inference_status = "stopped"

    services.append(ServiceInfo(
        name="vLLM/SGLang",
        port=settings.inference_port,
        internal_port=settings.inference_port,
        protocol="http",
        status=inference_status,
        description="Inference backend (vLLM or SGLang)"
    ))

    # LiteLLM
    litellm_status = "unknown"
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            master_key = os.environ.get("LITELLM_MASTER_KEY", "sk-master")
            headers = {"Authorization": f"Bearer {master_key}"}
            r = await client.get("http://localhost:4100/health", headers=headers)
            if r.status_code == 200:
                data = r.json()
                healthy_count = data.get("healthy_count", 0)
                litellm_status = "running" if healthy_count > 0 else "degraded"
            else:
                litellm_status = "error"
    except Exception:
        litellm_status = "stopped"

    services.append(ServiceInfo(
        name="LiteLLM",
        port=4100,
        internal_port=4000,
        protocol="http",
        status=litellm_status,
        description="API gateway and load balancer"
    ))

    # PostgreSQL
    is_postgres_reachable = await check_service("localhost", 5432)
    services.append(ServiceInfo(
        name="PostgreSQL",
        port=5432,
        internal_port=5432,
        protocol="tcp",
        status="running" if is_postgres_reachable else "stopped",
        description="Database for LiteLLM"
    ))

    # Redis
    is_redis_reachable = await check_service("localhost", 6379)
    services.append(ServiceInfo(
        name="Redis",
        port=6379,
        internal_port=6379,
        protocol="tcp",
        status="running" if is_redis_reachable else "stopped",
        description="Cache and rate limiting"
    ))

    # Prometheus
    prometheus_status = "unknown"
    try:
        async with httpx.AsyncClient(timeout=2) as client:
            r = await client.get("http://localhost:9090/-/healthy")
            prometheus_status = "running" if r.status_code == 200 else "error"
    except Exception:
        prometheus_status = "stopped"
    services.append(ServiceInfo(
        name="Prometheus",
        port=9090,
        internal_port=9090,
        protocol="http",
        status=prometheus_status,
        description="Metrics collection"
    ))

    # Grafana
    grafana_status = "unknown"
    try:
        async with httpx.AsyncClient(timeout=2) as client:
            r = await client.get("http://localhost:3001/api/health")
            grafana_status = "running" if r.status_code == 200 else "error"
    except Exception:
        grafana_status = "stopped"
    services.append(ServiceInfo(
        name="Grafana",
        port=3001,
        internal_port=3000,
        protocol="http",
        status=grafana_status,
        description="Metrics dashboards"
    ))

    # Frontend
    is_frontend_reachable = await check_service("localhost", 3000)
    services.append(ServiceInfo(
        name="Frontend",
        port=3000,
        internal_port=3000,
        protocol="http",
        status="running" if is_frontend_reachable else "stopped",
        description="Next.js web UI"
    ))

    config = SystemConfig(
        host=settings.host,
        port=settings.port,
        inference_port=settings.inference_port,
        api_key_configured=settings.api_key is not None,
        models_dir=str(settings.models_dir),
        data_dir=str(settings.data_dir),
        db_path=str(settings.db_path),
        sglang_python=settings.sglang_python,
        tabby_api_dir=settings.tabby_api_dir,
    )

    environment = EnvironmentInfo(
        controller_url=f"http://{hostname}:{settings.port}",
        inference_url=f"http://{hostname}:{settings.inference_port}",
        litellm_url=f"http://{hostname}:4100",
        frontend_url=f"http://{hostname}:3000",
    )

    return SystemConfigResponse(
        config=config,
        services=services,
        environment=environment
    )

File: controller/src/routes/system.ts
// CRITICAL
import type { Hono } from "hono";
import { connect } from "node:net";
import { hostname } from "node:os";
import { existsSync, readFileSync } from "node:fs";
import { join, resolve, sep } from "node:path";
import type { AppContext } from "../types/context";
import type { HealthResponse, SystemConfigResponse } from "../types/models";
import { badRequest, notFound } from "../core/errors";
import { estimateWeightsSizeBytes } from "../services/model-browser";
import { getGpuInfo } from "../services/gpu";
import { getSystemRuntimeInfo } from "../services/runtime-info";

/**
 * Register system routes.
 * @param app - Hono application.
 * @param context - App context.
 */
export const registerSystemRoutes = (app: Hono, context: AppContext): void => {
  /**
   * Check if a TCP service is reachable.
   * @param host - Hostname.
   * @param port - Port number.
   * @param timeoutMs - Timeout in ms.
   * @returns Promise resolving to availability.
   */
  const checkService = (host: string, port: number, timeoutMs = 1000): Promise<boolean> => {
    return new Promise((resolve) => {
      const socket = connect(port, host);
      const timer = setTimeout(() => {
        socket.destroy();
        resolve(false);
      }, timeoutMs);
      socket.once("connect", () => {
        clearTimeout(timer);
        socket.end();
        resolve(true);
      });
      socket.once("error", () => {
        clearTimeout(timer);
        resolve(false);
      });
    });
  };

  app.get("/health", async (ctx) => {
    const current = await context.processManager.findInferenceProcess(context.config.inference_port);
    let inferenceReady = false;
    if (current) {
      try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), 5000);
        const response = await fetch(`http://localhost:${context.config.inference_port}/health`, {
          signal: controller.signal,
        });
        clearTimeout(timeout);
        inferenceReady = response.status === 200;
      } catch {
        inferenceReady = false;
      }
    }

    const payload: HealthResponse = {
      status: "ok",
      version: "0.3.1",
      inference_ready: inferenceReady,
      backend_reachable: inferenceReady,
      running_model: current ? (current.served_model_name ?? current.model_path ?? null) : null,
    };
    return ctx.json(payload);
  });

  app.get("/status", async (ctx) => {
    const current = await context.processManager.findInferenceProcess(context.config.inference_port);
    return ctx.json({
      running: Boolean(current),
      process: current,
      inference_port: context.config.inference_port,
      launching: context.launchState.getLaunchingRecipeId(),
    });
  });

  app.get("/gpus", async (ctx) => {
    const gpus = getGpuInfo();
    return ctx.json({
      count: gpus.length,
      gpus,
    });
  });

  app.post("/vram-calculator", async (ctx) => {
    const body = await ctx.req.json().catch(() => ({}));
    if (!body || typeof body !== "object") {
      throw badRequest("Invalid payload");
    }

    const model = typeof body["model"] === "string" ? body["model"].trim() : "";
    const contextLength = Number(body["context_length"] ?? 0);
    const tpSize = Number(body["tp_size"] ?? 1);
    const kvDtype = typeof body["kv_dtype"] === "string" ? body["kv_dtype"] : "auto";

    if (!model) {
      throw badRequest("model is required");
    }
    if (!Number.isFinite(contextLength) || contextLength <= 0) {
      throw badRequest("context_length must be a positive number");
    }
    if (!Number.isFinite(tpSize) || tpSize <= 0) {
      throw badRequest("tp_size must be a positive number");
    }

    const resolved = resolve(model);
    const modelsRoot = resolve(context.config.models_dir);
    const rootPrefix = modelsRoot.endsWith(sep) ? modelsRoot : modelsRoot + sep;
    if (!resolved.startsWith(rootPrefix)) {
      throw badRequest("model must be inside models_dir");
    }
    if (!existsSync(resolved)) {
      throw notFound("Model path not found");
    }

    const weightsBytes = estimateWeightsSizeBytes(resolved, false);
    if (!weightsBytes || weightsBytes <= 0) {
      throw notFound("Model weights not found");
    }

    let config: Record<string, unknown> = {};
    const configPath = join(resolved, "config.json");
    if (existsSync(configPath)) {
      try {
        const raw = readFileSync(configPath, "utf-8");
        config = JSON.parse(raw) as Record<string, unknown>;
      } catch {
        config = {};
      }
    }

    const getNumber = (value: unknown): number | undefined => {
      if (typeof value === "number" && Number.isFinite(value)) return value;
      if (typeof value === "string" && value.trim() && !Number.isNaN(Number(value))) {
        return Number(value);
      }
      return undefined;
    };

    const layerCount =
      getNumber(config["num_hidden_layers"]) ??
      getNumber(config["n_layer"]) ??
      getNumber(config["num_layers"]);
    const hiddenSize =
      getNumber(config["hidden_size"]) ??
      getNumber(config["n_embd"]) ??
      getNumber(config["d_model"]) ??
      getNumber(config["dim"]);
    const headCount =
      getNumber(config["num_attention_heads"]) ??
      getNumber(config["n_head"]) ??
      getNumber(config["num_heads"]);
    const keyValueHeadCount =
      getNumber(config["num_key_value_heads"]) ??
      getNumber(config["num_kv_heads"]) ??
      headCount;
    const headDim =
      getNumber(config["head_dim"]) ??
      (hiddenSize && headCount ? hiddenSize / headCount : undefined);

    const kvBytesPerValue = kvDtype.toLowerCase() === "fp8" ? 1 : 2;
    let kvCacheBytes = 0;
    if (layerCount && keyValueHeadCount && headDim) {
      kvCacheBytes = contextLength * layerCount * keyValueHeadCount * headDim * 2 * kvBytesPerValue;
    }

    const weightsTotalGb = weightsBytes / 1024 ** 3;
    const weightsPerGpuGb = weightsTotalGb / tpSize;
    const kvCachePerGpuGb = kvCacheBytes > 0 ? kvCacheBytes / 1024 ** 3 / tpSize : 0;
    const activationsPerGpuGb = Math.max(0.5, weightsPerGpuGb * 0.1);
    const overheadPerGpuGb = 2.0;
    const perGpuGb = weightsPerGpuGb + kvCachePerGpuGb + activationsPerGpuGb + overheadPerGpuGb;
    const totalGb = perGpuGb * tpSize;

    const gpus = getGpuInfo();
    let perGpuCapacityGb = 0;
    if (gpus.length >= tpSize && tpSize > 0) {
      const candidates = gpus.slice(0, tpSize).map((gpu) => {
        if (gpu.memory_total_mb) return gpu.memory_total_mb / 1024;
        return gpu.memory_total / 1024 ** 3;
      });
      perGpuCapacityGb = Math.min(...candidates);
    }

    const fits = perGpuCapacityGb > 0 ? perGpuGb <= perGpuCapacityGb : true;
    const utilizationPercent = perGpuCapacityGb > 0 ? (perGpuGb / perGpuCapacityGb) * 100 : 0;

    return ctx.json({
      model_size_gb: weightsTotalGb,
      context_memory_gb: kvCachePerGpuGb * tpSize,
      overhead_gb: overheadPerGpuGb,
      total_gb: totalGb,
      fits_in_vram: fits,
      fits,
      utilization_percent: utilizationPercent,
      breakdown: {
        model_weights_gb: weightsPerGpuGb,
        kv_cache_gb: kvCachePerGpuGb,
        activations_gb: activationsPerGpuGb,
        per_gpu_gb: perGpuGb,
        total_gb: totalGb,
      },
    });
  });

  app.get("/config", async (ctx) => {
    const services: Array<{ name: string; port: number; internal_port: number; protocol: string; status: string; description?: string | null }> = [];
    services.push({
      name: "Controller",
      port: context.config.port,
      internal_port: context.config.port,
      protocol: "http",
      status: "running",
      description: "FastAPI model lifecycle manager",
    });

    let inferenceStatus = "unknown";
    try {
      const current = await context.processManager.findInferenceProcess(context.config.inference_port);
      if (current) {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), 2000);
        const response = await fetch(`http://localhost:${context.config.inference_port}/health`, {
          signal: controller.signal,
        });
        clearTimeout(timeout);
        inferenceStatus = response.status === 200 ? "running" : "error";
      } else {
        inferenceStatus = "stopped";
      }
    } catch {
      inferenceStatus = "stopped";
    }

    services.push({
      name: "vLLM/SGLang",
      port: context.config.inference_port,
      internal_port: context.config.inference_port,
      protocol: "http",
      status: inferenceStatus,
      description: "Inference backend (vLLM, SGLang, or llama.cpp)",
    });

    let litellmStatus = "unknown";
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 10_000);
      const masterKey = process.env["LITELLM_MASTER_KEY"] ?? "sk-master";
      const response = await fetch("http://localhost:4100/health", {
        headers: { Authorization: `Bearer ${masterKey}` },
        signal: controller.signal,
      });
      clearTimeout(timeout);
      if (response.status === 200) {
        const data = (await response.json()) as { healthy_count?: number };
        litellmStatus = (data.healthy_count ?? 0) > 0 ? "running" : "degraded";
      } else {
        litellmStatus = "error";
      }
    } catch {
      litellmStatus = "stopped";
    }

    services.push({
      name: "LiteLLM",
      port: 4100,
      internal_port: 4000,
      protocol: "http",
      status: litellmStatus,
      description: "API gateway and load balancer",
    });

    const postgresReachable = await checkService("localhost", 5432);
    services.push({
      name: "PostgreSQL",
      port: 5432,
      internal_port: 5432,
      protocol: "tcp",
      status: postgresReachable ? "running" : "stopped",
      description: "Database for LiteLLM",
    });

    const redisReachable = await checkService("localhost", 6379);
    services.push({
      name: "Redis",
      port: 6379,
      internal_port: 6379,
      protocol: "tcp",
      status: redisReachable ? "running" : "stopped",
      description: "Cache and rate limiting",
    });

    let prometheusStatus = "unknown";
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 2000);
      const response = await fetch("http://localhost:9090/-/healthy", { signal: controller.signal });
      clearTimeout(timeout);
      prometheusStatus = response.status === 200 ? "running" : "error";
    } catch {
      prometheusStatus = "stopped";
    }
    services.push({
      name: "Prometheus",
      port: 9090,
      internal_port: 9090,
      protocol: "http",
      status: prometheusStatus,
      description: "Metrics collection",
    });

    const frontendReachable = await checkService("localhost", 3000);
    services.push({
      name: "Frontend",
      port: 3000,
      internal_port: 3000,
      protocol: "http",
      status: frontendReachable ? "running" : "stopped",
      description: "Next.js web UI",
    });

    const runtime = await getSystemRuntimeInfo(context.config);

    const payload: SystemConfigResponse = {
      config: {
        host: context.config.host,
        port: context.config.port,
        inference_port: context.config.inference_port,
        api_key_configured: Boolean(context.config.api_key),
        models_dir: context.config.models_dir,
        data_dir: context.config.data_dir,
        db_path: context.config.db_path,
        sglang_python: context.config.sglang_python ?? null,
        tabby_api_dir: context.config.tabby_api_dir ?? null,
        llama_bin: context.config.llama_bin ?? null,
      },
      services,
      environment: {
        controller_url: `http://${hostname()}:${context.config.port}`,
        inference_url: `http://${hostname()}:${context.config.inference_port}`,
        litellm_url: `http://${hostname()}:4100`,
        frontend_url: `http://${hostname()}:3000`,
      },
      runtime,
    };

    return ctx.json(payload);
  });
};
